### DISCUSSION TITLE
getting "the trait `Sync` is not implemented for `(dyn wgpu::context::DynContext + 'static)"

### DISCUSSION THREAD

**Discussion by Umenokin** [gfx-rs/wgpu#4658] [Q&A] [answered]
I am trying to use wgpu with the shipyard.
I created `Unique` component which right now only has wgpu surface, device, and queue:
```
#[derive(Unique)]
pub struct Render {
    pub(crate) surface: wgpu::Surface,
    pub(crate) device: wgpu::Device,
    pub(crate) queue: wgpu::Queue,
}

...

app.world.add_unique(Render {
        surface,
        device,
        queue
 });
```

As a result, I am getting an error:
```
`(dyn wgpu::context::DynContext + 'static)` cannot be shared between threads safely
the trait `Sync` is not implemented for `(dyn wgpu::context::DynContext + 'static)`
required for `Arc<(dyn wgpu::context::DynContext + 'static)>` to implement `Send`
```

The cause of the error is having any of those three in my Render struct - `surface`, `device`, `queue`. But when I check the wgpu source code, I can see:

`static_assertions::assert_impl_all!(Instance: Send, Sync);`
`static_assertions::assert_impl_all!(Device: Send, Sync);`
`static_assertions::assert_impl_all!(Queue: Send, Sync);`

which I assume should mean that those are indeed has to be `Send` and `Sync`. More to that, it seems shipyard has the demo that doing exactly the same but does not suffer from that issue:

```
world.add_unique(Graphics {
        surface,
        device,
        queue,
        config,
        size,
        render_pipeline,
        vertex_buffer,
        vertex_count: 0,
        instance_buffer,
        scale_factor: window.scale_factor() as f32,
        screen_size_buffer,
        screen_size_bind_group,
        bunny_diffuse_bind_group,
        context,
        input: egui::RawInput::default(),
        egui_render_pass,
 });
```

Thanks! 🙏

**Comment by cwfitzgerald** [comment]
Are you running on the web? Wgpu types are only send and sync on native, on web they are !Send !Sync

**Reply by Umenokin** [reply]
Thank you @cwfitzgerald I thought it was likely the case but I could see any shipyard features in their demo working. But now I know I need to ask in shipyard's forum. Thanks!

**Reply by cwfitzgerald** [reply, marked as answer]
Send/Sync was only removed in 0.17 I think, so if they're still on 0.16,that could explain it

**Reply by Umenokin** [reply]
Thank you, @cwfitzgerald. You were right. It was a confusing combination of factors - shipyard using WGPU 0.15 and requires dedicated `thread_local` feature and different method used for !Sync WGPU which was not visible without this feature.

Thanks again!
